components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m

```diff
@@ -0,0 +1,657 @@
+  }
+}
+
+- (CGFloat)transitionPercentageForContentOffset:(CGPoint)contentOffset
+                                         offset:(CGFloat)offset
+                                       distance:(CGFloat)distance {
+  return 1.f - MAX(0.f, MIN(1.f, (self.transitionCompleteContentOffset - contentOffset.y - offset) /
+                                     distance));
+}
+
+- (CGFloat)midAnimationScrollToPositionForOffset:(CGPoint)targetContentOffset {
+  if (!self.contentScrollsToReveal) {
+    return NSNotFound;
+  }
+
+  CGFloat headerAnimationDistance = self.headerAnimationDistance;
+  CGFloat headerTransitionToTop =
+      [self transitionPercentageForContentOffset:targetContentOffset
+                                          offset:0
+                                        distance:headerAnimationDistance];
+  if (headerTransitionToTop >= FLT_EPSILON && headerTransitionToTop < 1.f) {
+    CGFloat contentHeaderFullyCoversTopHeaderContentOffset = self.transitionCompleteContentOffset;
+    CGFloat contentHeaderReachesTopHeaderContentOffset =
+        contentHeaderFullyCoversTopHeaderContentOffset - headerAnimationDistance;
+    return self.scrollViewIsDraggedToBottom ? contentHeaderReachesTopHeaderContentOffset
+                                            : contentHeaderFullyCoversTopHeaderContentOffset;
+  }
+
+  return NSNotFound;
+}
+
+@end
+
+#pragma mark - MDCBottomDrawerContainerViewController + Layout Values
+
+@implementation MDCBottomDrawerContainerViewController (LayoutValues)
+
+- (CGRect)presentingViewBounds {
+  return self.originalPresentingViewController.view.bounds;
+}
+
+- (BOOL)contentReachesFullscreen {
+  return self.contentHeightSurplus >= self.contentHeaderTopInset;
+}
+
+- (BOOL)contentScrollsToReveal {
+  return self.contentHeightSurplus > FLT_EPSILON;
+}
+
+- (CGFloat)topHeaderHeight {
+  if (!self.headerViewController) {
+    return 0.f;
+  }
+  CGFloat headerHeight = self.headerViewController.preferredContentSize.height;
+  return headerHeight + MDCDeviceTopSafeAreaInset();
+}
+
+- (CGFloat)contentHeaderHeight {
+  if (!self.headerViewController) {
+    return 0.f;
+  }
+  return self.headerViewController.preferredContentSize.height;
+}
+
+- (CGFloat)transitionCompleteContentOffset {
+  if (self.contentReachesFullscreen) {
+    CGFloat transitionCompleteContentOffset = self.contentHeaderTopInset;
+    transitionCompleteContentOffset -= self.topHeaderHeight - self.contentHeaderHeight;
+    return transitionCompleteContentOffset;
+  } else {
+    return self.contentHeightSurplus;
+  }
+}
+
+- (CGFloat)headerAnimationDistance {
+  CGFloat headerAnimationDistance =
+      kMDCBottomDrawerHeaderAnimationDistanceAddedDistanceFromTopSafeAreaInset;
+  if (self.contentReachesFullscreen) {
+    headerAnimationDistance += MDCDeviceTopSafeAreaInset();
+  }
+  return headerAnimationDistance;
+}
+
+- (CGFloat)addedContentHeightThreshold {
+  return MDCDeviceTopSafeAreaInset();
```

Comment:
Please add a TODO to replace this call with a call to safeAreaInsets.

Reply:
done!